### PR TITLE
Fix window function lowering for correlated subqueries

### LINKED ISSUE #9077: sql: Wrong result with window function, subquery
### What version of Materialize are you using?

From the shell:

```
v0.9.13-dev (0d9feb389)
```

### What was the issue?

If there is a window function inside a subquery, rows that should have matched the subquery are not returned.

### Is the issue reproducible? If so, please provide reproduction instructions.

```
DROP TABLE IF EXISTS t1;
DROP TABLE IF EXISTS t2;
CREATE TABLE t1 (f1 INTEGER);
INSERT INTO t1 VALUES (1), (2);
CREATE TABLE t2 (f1 INTEGER);
INSERT INTO t2 VALUES (1), (2);

SELECT *
  FROM t1
 WHERE f1 IN (
        SELECT ROW_NUMBER() OVER ()
          FROM t2
       );

 f1 
----
  1 
  2 <- Materialize fails to return this row 
(2 rows)
```

Materialize fails to return a row where `t1.f1 = 2` even though there is a row in the subquery for which `ROW_NUMBER` returns 2 in both products.

### PR DESCRIPTION
**Pull request by asenac**
### Motivation

Fixes #9077.

The reduction that computes the window function must be keyed on the columns from the outer context.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Review by mjibson — APPROVED** [review]

File: src/sql/src/plan/lowering.rs [hunk 1]
```diff
@@ -722,7 +722,12 @@ impl HirScalarExpr {
                                 // Record input arity here so that any group_keys that need to mutate get_inner
                                 // don't add those columns to the aggregate input.
                                 let input_arity = get_inner.typ().arity();
-                                let mut group_key = Vec::new();
+                                let mut group_key = col_map
```
Comment (wangandi) [inline src/sql/src/plan/lowering.rs:725]: > The reduction that computes the window function must be keyed on the columns from the outer context.
Can you add this as a comment here to explain what is happening? 

Why is it necessary to group by the columns from the outer context?

Comment (asenac) [reply src/sql/src/plan/lowering.rs:729]: Good point. I took it for granted but it's not obvious at all. Just added a comment explaining why.

**Comment by asenac** [comment]
> Does this PR have tests for what happens if there are more than one outer context?

It doesn't, but I think the fact that the outer context is the concatenation of all the elements from the outer query blocks the query sub-tree is correlated with is a contract in the lowering code we can rely on.

**Review by wangandi — APPROVED** [review]

**Comment by philip-stoev** [comment]
I need to write a new RQG grammar because the one that I have only rarely produces fully-valid queries with window functions. So please give me another day.

**Review by philip-stoev — APPROVED** [review]
I am now able to readily hit the original issue and this branch fixes it. No other wrong results or panics.

**Pull request merged**